frugen.c

```diff
@@ -816,120 +939,245 @@ int main(int argc, char *argv[])
+			json_object_add_with_type(section, "serial", chassis.serial.val, chassis.serial.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = chassis.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "chassis", section);
+		}
 
-	fname = argv[optind];
-	debug(1, "FRU info data will be stored in %s", fname);
+		if (has_product) {
+			section = json_object_new_object();
+			temp_obj = json_object_new_int(product.lang);
+			json_object_object_add(section, "lang", temp_obj);
+			json_object_add_with_type(section, "mfg", product.mfg.val, product.mfg.type);
+			json_object_add_with_type(section, "pname", product.pname.val, product.pname.type);
+			json_object_add_with_type(section, "serial", product.serial.val, product.serial.type);
+			json_object_add_with_type(section, "pn", product.pn.val, product.pn.type);
+			json_object_add_with_type(section, "ver", product.ver.val, product.ver.type);
+			json_object_add_with_type(section, "atag", product.atag.val, product.atag.type);
+			json_object_add_with_type(section, "file", product.file.val, product.file.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = product.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "product", section);
+		}
 
-	if (has_internal) {
-		debug(1, "FRU file will have an internal use area");
-		/* Nothing to do here, added for uniformity, the actual
-		 * internal use area can only be initialized from file */
-	}
+		if (has_board) {
+			section = json_object_new_object();
+			temp_obj = json_object_new_int(board.lang);
+			json_object_object_add(section, "lang", temp_obj);
+			json_object_add_with_type(section, "time", timebuf, 0);
+			json_object_add_with_type(section, "mfg", board.mfg.val, board.mfg.type);
+			json_object_add_with_type(section, "pname", board.pname.val, board.pname.type);
+			json_object_add_with_type(section, "serial", board.serial.val, board.serial.type);
+			json_object_add_with_type(section, "pn", board.pn.val, board.pn.type);
+			json_object_add_with_type(section, "file", board.file.val, board.file.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = board.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "board", section);
+		}
+		// dump to stdout
+		json_object_to_fd(fileno(stdout), json_root, 0);
+		json_object_put(json_root);
```

Comment:
This approach is more limiting because it does not allow decoding binary into JSON. I, for one, need exactly this functionality.

Reply:
Why? If you add a dump option, then it will dump to JSON the FRU constructed from any source, just like you need. If the dump option is implemented as specifying `-` for the output file, you won't even need a separate dry-run option. For your task you would just run `frugen -rz template.bin -`. Am I missing anything?

Reply:
So, just to clear, how exactly input options should be handled?

This way
|Command|Input|Output|
|:--|:--|:--|
|`./frugen --json --from-file template.json output.bin`|JSON|Bin|
|`./frugen --raw --from-file template.bin output.bin`|Bin|Bin|
|`./frugen --json --from-file template.json -`|JSON|Bin(stdout)|
|`./frugen --raw --from-file template.bin -`|Bin|Bin(stdout)|
|`./frugen --json --from-file template.json --dump output.json`|JSON|JSON|
|`./frugen --raw --from-file template.bin --dump output.json`|Bin|JSON|
|`./frugen --json --from-file template.json --dump -`|JSON|JSON(stdout)|
|`./frugen --raw --from-file template.bin --dump -`|Bin|JSON(stdout)|

Or this way
|Command|Input|Output|
|:--|:--|:--|
|`./frugen --json --from-file template.json output.bin`|JSON|Bin|
|`./frugen --raw --from-file template.bin output.bin`|Bin|Bin|
|`./frugen --json --from-file template.json -`|JSON|JSON(stdout)|
|`./frugen --raw --from-file template.bin -`|Bin|JSON(stdout)|

Reply:
> Do you have any objections?

No, I just thought it would be better to agree on this before coding.